### DISCUSSION TITLE
`Thrust / Cub` Codebase Structure

### DISCUSSION THREAD

**Discussion by jeromeku** [NVIDIA/cccl#1113] [Thrust]
Trying to understand the architecture of `Thrust` and `Cub`.  Is there documentation on the structure of the codebase, specifically, how the high level abstractions provided by `Thrust` and `Cub` are ultimately transformed into primitive `cuda` memory allocations / kernel launches?  Would appreciate any explanations or pointers.

Thanks!

**Comment by pauleonix** [comment]
For Thrust you can find most of the interesting CUDA-specific stuff in [`thrust/system/cuda/detail`](https://github.com/NVIDIA/cccl/tree/main/thrust/thrust/system/cuda/detail).

**Comment by jeromeku** [comment]
Thanks -- I'm looking more for an overview of how the code is organized, and an explanation of the core abstractions (e.g., `agent`) used to implement the various features of `thrust` and `cub`.

**Comment by gevtushenko** [comment]
Hello @jeromeku! 

For CUB you can take a look at the *Device-Scope* section of the [developer overview](https://github.com/NVIDIA/cccl/blob/main/cub/docs/developer_overview.rst). It should answer your questions on code organization surrounding CUB kernels. 

I don't think we have a developer overview for Thrust, but we definitely should have one. Could you please file a documentation issue, so we do not forget to write developer overview?

**Reply by jeromeku** [reply]
@gevtushenko 

Thanks! 

Filed here #1119.
